plugins/org.eclipse.cbi.p2repo.aggregator.engine.maven/src/org/eclipse/cbi/p2repo/aggregator/engine/maven/InstallableUnitMapping.java

```diff
@@ -588,38 +589,47 @@ public MavenItem map() throws CoreException {
 		if(mapped != null)
 			return mapped;
 
-		mapped = map(getId(), mappings);
-		if (mapped != null && this.versionFormat != VersionFormat.MAVEN_SNAPSHOT) {
-			if (mapped.getMavenMapping().isSnapshot()) {
-				versionFormat = VersionFormat.MAVEN_SNAPSHOT; // override global format
-			} else { // fade out the following workaround?
-				String mappedVersion = this.mapped.getMappedVersion();
-				if (mappedVersion != null && mappedVersion.contains(VersionUtil.DASH_SNAPSHOT))
+		// first lookup defined mappings
+		Optional<MavenItem> staticMapping = mappings.stream().map(mapping -> mapping.map(getId(), getVersion())) //
+				.filter(Objects::nonNull)
+				.findFirst();
+		if (staticMapping.isPresent()) {
+			mapped = staticMapping.get();
+			if (this.versionFormat != VersionFormat.MAVEN_SNAPSHOT) {
+				if (mapped.getMavenMapping().isSnapshot()) {
 					versionFormat = VersionFormat.MAVEN_SNAPSHOT; // override global format
+				} else { // fade out the following workaround?
+					String mappedVersion = mapped.getMappedVersion();
+					if (mappedVersion != null && mappedVersion.contains(VersionUtil.DASH_SNAPSHOT))
+						versionFormat = VersionFormat.MAVEN_SNAPSHOT; // override global format
+				}
 			}
+			return mapped;
 		}
-		return mapped;
-	}
 
-	private MavenItem map(String id, List<MavenMapping> mappings) throws InvalidParameterException {
-		MavenItem item = null;
+		// TODO: lookup <file>/META-INF/maven/pom.xml
+
+		// read from p2 properties as generated by Tycho
+		String p2GroupIdProperty = installableUnit.getProperty("maven-groupId");
+		String p2ArtifactIdProperty = installableUnit.getProperty("maven-artifactId");
+		String p2VersionProperty = installableUnit.getProperty("maven-version");
+		if (p2GroupIdProperty != null && p2ArtifactIdProperty != null) {
+			mapped = AggregatorFactory.eINSTANCE.createMavenItem();
+			mapped.setGroupId(p2GroupIdProperty);
+			mapped.setArtifactId(p2ArtifactIdProperty);
+			mapped.setMappedVersion(p2VersionProperty); // TODO handle snapshots
+		}
 
-		for(MavenMapping mapping : mappings)
-			if((item = mapping.map(id, getVersion())) != null)
-				return item;
-
-		StringBuilder mappingDescriptor = new StringBuilder();
-		boolean first = true;
-		for(MavenMapping mapping : mappings) {
-			if(first)
-				first = false;
-			else
-				mappingDescriptor.append(',');
-			mappingDescriptor.append(mapping.toString());
+		if (mapped == null) {
+			throw new CoreException(
+					new Status(IStatus.ERROR, getClass(), "Could not generate Maven mapping for " + getId()));
 		}
+		return mapped;
+	}
 
-		throw new InvalidParameterException(String.format("Unable to map IU to maven artifact: id=%s, mappings=%s", id,
-				mappingDescriptor.toString()));
+	private InstallableUnitMapping resolveMapping(IRequiredCapability requirement) {
```

Comment:
This is going to result in n^2 performance.

Reply:
I'm fine merging it as soon as we have a good way to lower priority of MavenMapping.DEFAULT_MAPPING